src/meta-srv/src/election/mysql.rs

```diff
@@ -0,0 +1,836 @@
+    }
+
+    fn insert_once(&self) -> String {
+        format!(
+            "INSERT IGNORE INTO {} (k, v) VALUES ('__place_holder_for_lock', '');",
+            self.table_name
+        )
+    }
+
+    fn campaign_sql(&self) -> String {
+        format!("SELECT * FROM {} FOR UPDATE;", self.table_name)
+    }
+
+    fn put_value_with_lease_sql(&self) -> String {
+        format!(
+            r#"
+            INSERT INTO {} (k, v) VALUES (
+                ?,
+                CONCAT(
+                    ?,
+                    '{}',
+                    DATE_FORMAT(DATE_ADD(NOW(4), INTERVAL ? SECOND), '%Y-%m-%d %T.%f')
+                )
+            )
+            ON DUPLICATE KEY UPDATE v = VALUES(v);
+            "#,
+            self.table_name, LEASE_SEP
+        )
+    }
+
+    fn update_value_with_lease_sql(&self) -> String {
+        format!(
+            r#"UPDATE {}
+               SET v = CONCAT(?, '{}', DATE_FORMAT(DATE_ADD(NOW(4), INTERVAL ? SECOND), '%Y-%m-%d %T.%f'))
+               WHERE k = ? AND v = ?"#,
+            self.table_name, LEASE_SEP
+        )
+    }
+
+    fn get_value_with_lease_sql(&self) -> String {
+        format!(
+            r#"SELECT v, DATE_FORMAT(NOW(4), '%Y-%m-%d %T.%f') FROM {} WHERE k = ?"#,
+            self.table_name
+        )
+    }
+
+    fn get_value_with_lease_by_prefix_sql(&self) -> String {
+        format!(
+            r#"SELECT v, DATE_FORMAT(NOW(4), '%Y-%m-%d %T.%f') FROM {} WHERE k LIKE ?"#,
+            self.table_name
+        )
+    }
+
+    fn delete_value_sql(&self) -> String {
+        format!("DELETE FROM {} WHERE k = ?;", self.table_name)
+    }
+}
+
+/// Parse the value and expire time from the given string. The value should be in the format "value || LEASE_SEP || expire_time".
+fn parse_value_and_expire_time(value: &str) -> Result<(String, Timestamp)> {
+    let (value, expire_time) = value
+        .split(LEASE_SEP)
+        .collect_tuple()
+        .context(UnexpectedSnafu {
+            violated: format!(
+                "Invalid value {}, expect node info || {} || expire time",
+                value, LEASE_SEP
+            ),
+        })?;
+    // Given expire_time is in the format 'YYYY-MM-DD HH24:MI:SS.MS'
+    let expire_time = match Timestamp::from_str(expire_time, None) {
+        Ok(ts) => ts,
+        Err(_) => UnexpectedSnafu {
+            violated: format!("Invalid timestamp: {}", expire_time),
+        }
+        .fail()?,
+    };
+    Ok((value.to_string(), expire_time))
+}
+
+#[derive(Debug, Clone, Default)]
+struct MySqlLeaderKey {
+    name: Vec<u8>,
+    key: Vec<u8>,
+    rev: i64,
+    lease: i64,
+}
+
+impl LeaderKey for MySqlLeaderKey {
+    fn name(&self) -> &[u8] {
+        &self.name
+    }
+
+    fn key(&self) -> &[u8] {
+        &self.key
+    }
+
+    fn revision(&self) -> i64 {
+        self.rev
+    }
+
+    fn lease_id(&self) -> i64 {
+        self.lease
+    }
+}
+
+enum Executor<'a> {
+    Default(MutexGuard<'a, MySqlConnection>),
```

Comment:
Why not always use the txn? The election is racing against other metasrv process.

Reply:
Followers just make a read to check the lease, which does not need transactions.